src/awkward/operations/ak_firsts.py

```diff
@@ -44,24 +44,55 @@ def firsts(array, axis=1, *, highlevel=True, behavior=None):
 
 
 def _impl(array, axis, highlevel, behavior):
-    layout = ak.operations.to_layout(array, allow_record=False, allow_other=False)
+    layout = ak.operations.to_layout(array)
+    behavior = ak._util.behavior_of(array, behavior=behavior)
     posaxis = layout.axis_wrap_if_negative(axis)
 
+    if not ak._util.is_integer(axis):
+        raise ak._errors.wrap_error(
+            TypeError(f"'axis' must be an integer, not {axis!r}")
+        )
+
     if posaxis == 0:
-        if len(layout) == 0:
-            out = None
+        # specialized logic; it's tested in test_0582-propagate-context-in-broadcast_and_apply.py
+        slicer = ak.from_iter([None, 0])
+        if layout.length == 0:
```

Comment:
This is clever.

Perhaps a comment like this (but not this):
```suggestion
        # Build an integer-typed slice array, so that we can
        # ensure we have advanced indexing for both length==0
        # and length > 0 cases.
        slicer = ak.from_iter([None, 0])
        if layout.length == 0:
```